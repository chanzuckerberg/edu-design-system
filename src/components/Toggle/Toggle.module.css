@import '../../design-tokens/mixins.css';

/**
 * The interactive pill shaped component with a circle to indicate toggle status.
 */
.toggle__button {
  /* The transparent border is used to support Windows high contrast mode and reduce potential jiggling with adding / removing borders */
  border: calc(var(--eds-theme-border-width) * 1px) solid transparent;
  padding: 0;

  position: relative;
  display: inline-flex;

  height: calc(var(--eds-spacing-size-3) * 1px);
  width: calc(var(--eds-spacing-size-6) * 1px);

  flex-shrink: 0;
  align-items: center;

  border-radius: calc(var(--eds-border-radius-full) * 1px);

  cursor: pointer;

  transition: background-color calc(var(--eds-anim-move-quick) * 1s) ease-in-out;

  @media screen and (prefers-reduced-motion) {
    transition: none;
  }
}

/**
 * Focus state of the toggle button.
 */
.toggle__button:focus-visible {
  outline: calc(var(--eds-spacing-size-quarter) * 1px) solid var(--eds-theme-color-border-utility-focus);
  outline-offset: calc(var(--eds-spacing-size-quarter) * 1px);
}

@supports not selector(:focus-visible) {
  .toggle__button:focus {
    outline: calc(var(--eds-spacing-size-quarter) * 1px) solid var(--eds-theme-color-border-utility-focus);
    outline-offset: calc(var(--eds-spacing-size-quarter) * 1px);
  }
}

/**
 * Unchecked state of the toggle button.
 */
.toggle__button--unchecked {
  background-color: var(--eds-theme-color-background-utility-default-medium-emphasis);
}

/**
 * Checked state of the toggle button.
 */
.toggle__button--checked {
  background-color: var(--eds-theme-color-background-utility-default-high-emphasis);
}

/**
 * Disabled state of the toggle button.
 */
.toggle__button:disabled {
  background-color: var(--eds-theme-color-background-utility-disabled-medium-emphasis);
  cursor: not-allowed;
}

/**
 * Disabled state of the associated label text to the toggle.
 */
.toggle__button:disabled ~ .toggle__label {
  color: var(--eds-theme-color-text-utility-disabled-primary);
  cursor: not-allowed;
}

/**
 * The (sub-)label text to name the toggle.
 */
.toggle__label {
  margin-bottom: 0;

  color: var(--eds-theme-color-text-utility-default-primary);
}

.toggle__sublabel {
  display: block;
  color: var(--eds-theme-color-text-utility-default-secondary);
}

.toggle__button:disabled ~ .toggle__sublabel {
  color: var(--eds-theme-color-text-utility-disabled-primary);
  cursor: not-allowed;
}

/**
 * Wraps the toggle button and the (sub-)label text.
 */
.toggle__wrapper {
  display: flex;
  align-items: center;

  gap: calc(var(--eds-spacing-size-1) * 1px);
}

.toggle--has-sublabel {
  align-items: flex-start;
}

/**
 * The circle indicator of the toggle button.
 */
.toggle__button-thumb {
  display: inline-block;

  height: calc(var(--eds-spacing-size-2-and-half) * 1px);
  width: calc(var(--eds-spacing-size-2-and-half) * 1px);

  /* The transparent border is used to support Windows high contrast mode and reduce potential jiggling with adding / removing borders */
  border: calc(var(--eds-theme-border-width) * 1px) solid transparent;
  border-radius: calc(var(--eds-border-radius-full) * 1px);

  background-color: var(--eds-theme-color-background-utility-base-1);

  transition: transform calc(var(--eds-anim-move-quick) * 1s) ease-in-out;

  @media screen and (prefers-reduced-motion) {
    transition: none;
  }
}

/**
 * The unchecked state of the circle indicator.
 */
.toggle__button--unchecked > .toggle__button-thumb {
  transform: translateX(calc(0% + var(--eds-theme-border-width) * 1px));
}

/**
 * The checked state of the circle indicator.
 */
.toggle__button--checked > .toggle__button-thumb {
  transform: translateX(calc((var(--eds-theme-border-width) * 1px) + (var(--eds-spacing-size-3) * 1px)));
}