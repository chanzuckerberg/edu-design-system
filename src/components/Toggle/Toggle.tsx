import { Switch, Label, Field, Description } from '@headlessui/react';
import clsx from 'clsx';
import React from 'react';
import type { ReactNode } from 'react';
import type { EitherInclusive, ExtractProps } from '../../util/utility-types';
import { default as EDSLabel } from '../Label';
import styles from './Toggle.module.css';

type ToggleLabelProps = {
  /**
   * Visible text label for the toggle.
   */
  children: ReactNode;
  /**
   * Additional classnames passed in for styling the toggle label.
   */
  className?: string;
};

type ToggleButtonProps = {
  /**
   * Additional classnames passed in for styling the toggle button.
   */
  className?: string;
  /**
   * Whether toggle is checked or unchecked.
   */
  checked: boolean;
  /**
   * Callback called when toggle state is changed.
   */
  onChange: (checked: boolean) => void;
  /**
   * Whether toggle is disabled.
   */
  disabled?: boolean;
};

type ToggleProps = ToggleButtonProps & {
  children?: ReactNode;
  sublabel?: ReactNode;
} & EitherInclusive<
    {
      /**
       * Visible text label for the component.
       */
      label: ReactNode;
    },
    {
      /**
       * Aria-label to provide an accesible name for the text input if no visible label is provided.
       */
      'aria-label': string;
    }
  >;

const ToggleLabel = ({ children, className }: ToggleLabelProps) => {
  const componentClassName = clsx(styles['toggle__label'], className);
  return <Label as={EDSLabel} className={componentClassName} text={children} />;
};

const ToggleSubLabel = ({ children, className }: ToggleLabelProps) => {
  const componentClassName = clsx(styles['toggle__sublabel'], className);
  return (
    <Description as="span" className={componentClassName}>
      {children}
    </Description>
  );
};

const ToggleButton = ({ className, checked, ...other }: ToggleButtonProps) => (
  <Switch
    checked={checked}
    className={clsx(
      styles['toggle__button'],
      checked
        ? styles['toggle__button--checked']
        : styles['toggle__button--unchecked'],
      className,
    )}
    {...other}
  >
    {/* This span is the circular part of the toggle */}
    <span className={styles['toggle__button-thumb']} />
  </Switch>
);

/**
 * In order to use the individual <Toggle.Label> and <Toggle.Button> sub-components for customization,
 * they need to be wrapped in this <ToggleWrapper> component for some functionality HeadlessUI provides.
 *
 * Instead of direct assignment to ToggleWrapper, wrapping into a custom component prevents TypeScript erroring of using private Headless Switch Group.
 */
const ToggleWrapper = (props: ExtractProps<typeof Field>) => (
  <Field {...props} />
);

/**
 * `import {Toggle} from "@chanzuckerberg/eds";`
 *
 * Toggle wrapping the Headless UI Switch component https://headlessui.dev/react/switch, generally used as an input for controlling between two states.
 *
 * **NOTE**: This component requires `label` or `aria-label` prop
 */
export const Toggle = ({ label, sublabel, ...other }: ToggleProps) => {
  const wrapperClassName = clsx(
    styles['toggle__wrapper'],
    sublabel && styles['toggle--has-sublabel'],
  );
  return label ? (
    <ToggleWrapper as="div" className={wrapperClassName}>
      <ToggleButton {...other} />
      <ToggleLabel>
        <>
          {label}
          {sublabel && <ToggleSubLabel>{sublabel}</ToggleSubLabel>}
        </>
      </ToggleLabel>
    </ToggleWrapper>
  ) : (
    <ToggleButton {...other} />
  );
};

Toggle.displayName = 'Toggle';
ToggleLabel.displayName = 'Toggle.Label';
ToggleButton.displayName = 'Toggle.Button';
ToggleWrapper.displayName = 'Toggle.Wrapper';
ToggleSubLabel.displayName = 'Toggle.SubLabel';

Toggle.Label = ToggleLabel;
Toggle.Button = ToggleButton;
Toggle.Wrapper = ToggleWrapper;
Toggle.SubLabel = ToggleSubLabel;
