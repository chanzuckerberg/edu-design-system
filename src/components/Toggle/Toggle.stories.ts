import { BADGE } from '@geometricpanda/storybook-addon-badges';
import type { StoryObj, Meta } from '@storybook/react';
import type React from 'react';

import { Toggle } from './Toggle';

export default {
  title: 'PleaseUpdateThisToADifferentFolder/Toggle',
  component: Toggle,
  args: {
    label: 'Lorem ipsum',
    checked: false,
    onChange: () => {},
  },
  parameters: {
    badges: [BADGE.BETA],
  },
} as Meta<Args>;

type Args = React.ComponentProps<typeof Toggle>;

export const Default: StoryObj<Args> = {};
