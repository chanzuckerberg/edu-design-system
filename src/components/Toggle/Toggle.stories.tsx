import { BADGE } from '@geometricpanda/storybook-addon-badges';
import type { StoryObj, Meta } from '@storybook/react';
import React, { useState } from 'react';

import { Toggle } from './Toggle';

const InteractiveToggle = ({ checked = false, onChange, ...other }: Args) => {
  const [selected, setSelected] = useState(checked);
  return (
    <Toggle
      checked={selected}
      onChange={() => {
        setSelected(!selected);
        onChange(selected);
      }}
      {...other}
    />
  );
};

export default {
  title: 'Components/Toggle',
  component: Toggle,
  args: {
    label: 'Lorem ipsum',
    checked: false,
  },
  parameters: {
    layout: 'centered',
    badges: ['1.0', BADGE.BETA],
  },
  render: (args) => <InteractiveToggle {...args} />,
} as Meta<Args>;

type Args = React.ComponentProps<typeof Toggle>;

export const Default: StoryObj<Args> = {};

export const CheckedTrue: StoryObj<Args> = {
  args: {
    checked: true,
  },
};
export const Disabled: StoryObj<Args> = {
  args: {
    disabled: true,
  },
};
export const WithoutLabel: StoryObj<Args> = {
  args: {
    label: undefined,
    'aria-label': 'Lorem ipsum',
  },
};

export const CustomPositioning: StoryObj<Args> = {
  parameters: {
    docs: {
      source: {
        type: 'dynamic',
      },
    },
  },
  render: () => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const [selected, setSelected] = useState(false);
    return (
      <Toggle.Wrapper as="div" className="flex flex-col items-center">
        <Toggle.Label>Custom positioned label</Toggle.Label>
        <Toggle.Button
          checked={selected}
          onChange={() => {
            setSelected(!selected);
          }}
        />
      </Toggle.Wrapper>
    );
  },
};
