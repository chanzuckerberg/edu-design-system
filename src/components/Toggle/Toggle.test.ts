import { generateSnapshots } from '@chanzuckerberg/story-utils';
import * as stories from './Toggle.stories';

describe('<Toggle />', () => {
  generateSnapshots(stories);
});
