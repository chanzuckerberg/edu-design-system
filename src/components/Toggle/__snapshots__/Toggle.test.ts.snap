// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> CheckedTrue story renders snapshot 1`] = `
<div
  class="toggle__wrapper"
  data-headlessui-state=""
>
  <button
    aria-checked="true"
    aria-labelledby="headlessui-label-:r5:"
    class="toggle__button toggle__button--checked"
    data-checked=""
    data-headlessui-state="checked"
    id="headlessui-control-:r3:"
    role="switch"
    tabindex="0"
    type="button"
  >
    <span
      class="toggle__button-thumb"
    />
  </button>
  <label
    class="toggle__label"
    data-headlessui-state=""
    for="headlessui-control-:r3:"
    id="headlessui-label-:r5:"
  >
    <span
      class="text text--label-lg"
    >
      Lorem ipsum
    </span>
  </label>
  <span
    hidden=""
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`<Toggle /> CustomPositioning story renders snapshot 1`] = `
<div
  class="flex flex-col items-center"
  data-headlessui-state=""
>
  <label
    class="toggle__label"
    data-headlessui-state=""
    for="headlessui-control-:ra:"
    id="headlessui-label-:rb:"
  >
    <span
      class="text text--label-lg"
    >
      Custom positioned label
    </span>
  </label>
  <button
    aria-checked="false"
    aria-labelledby="headlessui-label-:rb:"
    class="toggle__button toggle__button--unchecked"
    data-headlessui-state=""
    id="headlessui-control-:ra:"
    role="switch"
    tabindex="0"
    type="button"
  >
    <span
      class="toggle__button-thumb"
    />
  </button>
  <span
    hidden=""
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`<Toggle /> Default story renders snapshot 1`] = `
<div
  class="toggle__wrapper"
  data-headlessui-state=""
>
  <button
    aria-checked="false"
    aria-labelledby="headlessui-label-:r2:"
    class="toggle__button toggle__button--unchecked"
    data-headlessui-state=""
    id="headlessui-control-:r0:"
    role="switch"
    tabindex="0"
    type="button"
  >
    <span
      class="toggle__button-thumb"
    />
  </button>
  <label
    class="toggle__label"
    data-headlessui-state=""
    for="headlessui-control-:r0:"
    id="headlessui-label-:r2:"
  >
    <span
      class="text text--label-lg"
    >
      Lorem ipsum
    </span>
  </label>
  <span
    hidden=""
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`<Toggle /> Disabled story renders snapshot 1`] = `
<div
  class="toggle__wrapper"
  data-headlessui-state=""
>
  <button
    aria-checked="false"
    aria-labelledby="headlessui-label-:r8:"
    class="toggle__button toggle__button--unchecked"
    data-disabled=""
    data-headlessui-state="disabled"
    disabled=""
    id="headlessui-control-:r6:"
    role="switch"
    tabindex="0"
    type="button"
  >
    <span
      class="toggle__button-thumb"
    />
  </button>
  <label
    class="toggle__label"
    data-headlessui-state=""
    for="headlessui-control-:r6:"
    id="headlessui-label-:r8:"
  >
    <span
      class="text text--label-lg"
    >
      Lorem ipsum
    </span>
  </label>
  <span
    hidden=""
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`<Toggle /> WithoutLabel story renders snapshot 1`] = `
<button
  aria-checked="false"
  aria-label="Lorem ipsum"
  class="toggle__button toggle__button--unchecked"
  data-headlessui-state=""
  id="headlessui-switch-:r9:"
  role="switch"
  tabindex="0"
  type="button"
>
  <span
    class="toggle__button-thumb"
  />
</button>
`;
